JOSS paper (#76)
OceanBioME/OceanBioME.jl: paper/paper.md

Reviewer: As a general comment  I didn't see any reference to the figures in the main text so as is now the figures just appear at some point? Wouldn't it be better if we reference the figures, if possible, where we describe the potential abilities of the package?

Author: This is a good point. I've added a mention of them but not sure I've done it in the best way

    affiliation: "1, 2"
  - name: Si Chen
    affiliation: "1, 2"
  - name: Gregory LeClaire Wagner
    orcid: 0000-0001-5317-2445
    affiliation: 3
  - name: Collaborators
affiliations:
 - name: Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge, United Kingdom
   index: 1
 - name: Centre for Climate Repair, Cambridge, United Kingdom
   index: 2
 - name: Massachusetts Institute of Technology
   index: 3
date: 15 March 2023
bibliography: paper.bib
---

# Summary

``OceanBioME.jl`` is a flexible modelling environment written in Julia [@julia] for modelling the coupled interactions between ocean biogeochemistry, carbonate chemistry, and physics.
OceanBioME can be used as a stand-alone box model, or integrated into ``Oceananigans.jl`` [@Oceananigans] simulations of ocean-flavoured fluid dynamics in one-, two-, or three-dimensions.
As a result, OceanBioME and Oceananigans can be used to simulate the biogeochemical response across an enormous range of scales: from surface boundary layer turbulence at the meter scale to eddying global ocean simulations at the planetary scale, and on computational systems ranging from laptops to supercomputers.

> An example small-scale problem is shown in Figure 1.

Reviewer: ```suggestion
An example small-scale problem is shown in \autoref{fig1}.
```

Reviewer: This sentence reads a bit as if the problem in figure 1 is small and not that as it involves small scale flow features. Perhaps rephrase to "An example of a problem involving kilimeter-lenght flow features is shown..."?

OceanBioME leverages Julia's multiple dispatch and effective inline capabilities to fuse its computations directly into existing Oceananigans kernels, thus maintaining Oceananigans' bespoke performance, memory- and cost-efficiency on GPUs in OceanBioME-augmented simulations.

![Here we replicate the Eady problem where a background buoyancy gradient and corresponding thermal wind generate a sub-mesoscale eddy, roughly following the setup of Taylor (2016).
To this physical setup, we added a medium complexity (9 tracers) biogeochemical model, some of which are shown above.
On top of this, we added particles modelling the growth of sugar kelp which are free-floating and advected by the flow, and carbon dioxide exchange from the air.
A key advantage of writing ``OceanBioME.jl`` in Julia is that it offers accessibility similar to high-level languages such as Python, with the speed of languages like C and Fortran and built-in parallelism.
This means that models can be run significantly faster than the equivalent in other high-level languages.

> ``OceanBioME.jl`` can run on GPUs, allowing the above model (1 km × 1 km × 100 m with 64 × 64 × 16 grid points) to simulate 10 days of evolution in about 30 minutes of computing time on an Nvidia P100 GPU.](eady_example.png)

Reviewer: ```suggestion
``OceanBioME.jl`` can run on GPUs, allowing the above model (1 km × 1 km × 100 m with 64 × 64 × 16 grid points) to simulate 10 days of evolution in about 30 minutes of computing time on an Nvidia P100 GPU. \label{fig1}](eady_example.png)
```

Reviewer: Caption of figure 1 seems to contain context beyond description of the figure. E.g.,

> A key advantage of writing ``OceanBioME.jl`` in Julia is that it offers accessibility similar to high-level languages such as Python, with the speed of languages like C and Fortran and built-in parallelism. This means that models can be run significantly faster than the equivalent in other high-level languages."

Do we want this as part of the caption or part of the main text?

Author: I think this is probably adequately described in the main text already on line 39, so should probably just be cut from the caption?

Reviewer: Yes, cut or shorten I'd say

OceanBioME is built with a highly modular design that allows user control and customization.
There are three distinct module types implemented in OceanBioME.jl.
First, tracer-based ecosystem modules are formulated as a set of coupled ordinary differential equations.
These equations can be solved by OceanBioME as box models, which is particularly useful for testing.
The same modules can be integrated by Oceananigans to provide tracer-based ecosystem models.
Second, boundary modules contain sets of equations which provide information at the top and bottom of the ocean.
For example, air-sea gas exchange modules calculate the flux of carbon dioxide and oxygen at the sea surface, while sediment modules calculate fluxes of carbon and oxygen at the seafloor.

> The third module type is "biologically active" particles.

Reviewer: suggestion: perhaps 3 bullet points will improve readability here?

This framework is made possible by our contributions to Oceananigans, adding a streamlined user interface to swap biogeochemical models with no modification to other model configurations.
This interface also facilitates rapid prototyping, as models can be coded to be in a much more accessible way, which is not possible with existing biogeochemical models.

OceanBioME was designed specifically to study ocean carbon dioxide removal (OCDR) strategies.
Assessing the effectiveness and impacts of OCDR is challenging due to the complexities of the interactions between the biological, chemical, and physical processes involved in the carbon cycle.
Moreover, field trials of OCDR interventions are generally small-scale and targeted, while the intervention required to have a climate-scale impact is regional or global.
We have built OceanBioME to meet these challenges by creating tools that provide a modular interface to the different components within the ocean modelling framework provided by Oceananigans.
This allows easy access to a suite of biogeochemical models ranging from simple idealized to full-complexity models.

> Figure 2 shows a simple column model with an OCDR intervention (macroalgae growth) added after a warm-up period, which increases the carbon export of the system.

Reviewer: ```suggestion
\autoref{fig2} shows a simple column model with an OCDR intervention (macroalgae growth) added after a warm-up period, which increases the carbon export of the system.
```

![Here we show the results of a 1D model, forced by idealised light and mixing, which qualitatively reproduces the biogeochemical cycles in the North Atlantic.
We then add kelp (500 frond / m² in the top 50 m of water) in December of the 3ʳᵈ year (black vertical line) which causes an increase in air-sea carbon dioxide exchange and sinking export, as well as a change in the phytoplankton growth cycle.

> Plot made with `Makie` [@makie].](column_example.png)

Reviewer: ```suggestion
Figure made with `Makie` [@makie]. \label{fig1}](column_example.png)
```

For example among the open-source alternatives NEMO [@nemo] provides a comprehensive global biogeochemical modelling framework but requires complex configuration and is unsuited for local ecosystem modelling, while MACMODS [@macmods] provides more limited functionality on a slower platform. -->

Finally, this software is currently facilitating multiple research projects into ocean CDR which would have been significantly harder with other solutions.
For example, Chen (In prep.) is using the active particle coupling provided to investigate the effects of location and planting density of kelp in the open ocean on their carbon drawdown effect, as in the example above.
Additionally, Strong-Wright (In prep.) is using the coupling of both the biogeochemistry and easy interface to couple the physics to study flow interactions with a fully resolved giant kelp forest model including the effects on nutrient transport and distribution.

# Acknowledgements

> We would like to thank the CLiMA team and ``Oceananigans`` contributors for their fantastic project. We are also very grateful for the support and funding of the [Centre for Climate Repair at Cambridge](https://www.climaterepair.cam.ac.uk/) and the [Gordon and Betty Moore Foundation](https://www.moore.org/).

Reviewer: ```suggestion
We would like to thank the [Climate Modeling Alliance](https://clima.caltech.edu) team and ``Oceananigans.jl`` contributors for their fantastic project. We are also very grateful for the support and funding of the [Centre for Climate Repair at Cambridge](https://www.climaterepair.cam.ac.uk/) and the [Gordon and Betty Moore Foundation](https://www.moore.org/).
```

---
title: 'OceanBioME.jl: A flexible environment for modelling the coupled interactions between ocean biogeochemistry and physics'
tags:
  - julia
  - biogeochemistry
  - climate
  - ocean
  - carbon
authors:
  - name: Jago Strong-Wright
    orcid: 0000-0002-7174-5283
    corresponding: true
    affiliation: "1, 2"
  - name: John R Taylor
    affiliation: "1, 2"
  - name: Si Chen
    affiliation: "1, 2"
  - name: Gregory LeClaire Wagner
    orcid: 0000-0001-5317-2445
    affiliation: 3
  - name: Navid C Constantinou

>     affiliation: 4

Reviewer: ```suggestion
    affiliation: 4, 5
```

affiliations:
 - name: Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge, United Kingdom
   index: 1
 - name: Centre for Climate Repair, Cambridge, United Kingdom
   index: 2
 - name: Massachusetts Institute of Technology
   index: 3
 - name: Australian National University

>    index: 4

Reviewer: ```suggestion
   index: 4
 - name: Australian Research Council Centre for Climate Extremes
   index: 5
```

>   - name: Navid C Constantinou
>     affiliation: 4, 5
>   - name: Simone Silvestri
>     affiliation: 3
>   - name: Gregory LeClaire Wagner
>     orcid: 0000-0001-5317-2445
>     affiliation: 3
>   - name: John R Taylor
>     affiliation: "1, 2"
> affiliations:
>  - name: Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge, United Kingdom
>    index: 1
>  - name: Centre for Climate Repair, Cambridge, United Kingdom
>    index: 2
>  - name: Massachusetts Institute of Technology
>    index: 3
>  - name: Australian National University
>    index: 4
>  - name: Australian Research Council Centre for Climate Extremes
>    index: 5

Reviewer: ```suggestion
  - name: Navid C Constantinou
    affiliation: 3, 4
  - name: Simone Silvestri
    affiliation: 5
  - name: Gregory LeClaire Wagner
    orcid: 0000-0001-5317-2445
    affiliation: 5
  - name: John R Taylor
    affiliation: "1, 2"
affiliations:
 - name: Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge, United Kingdom
   index: 1
 - name: Centre for Climate Repair, Cambridge, United Kingdom
   index: 2
 - name: Australian National University
   index: 3
 - name: Australian Research Council Centre for Climate Extremes
   index: 4
 - name: Massachusetts Institute of Technology
   index: 5
 - name: Massachusetts Institute of Technology
   index: 5
```

this is just so affiliations are numbered in the order they appear

> date: 15 March 2023

Reviewer: ```suggestion
date: 24 June 2023
```

bibliography: paper.bib
---

# Summary

``OceanBioME.jl`` is a flexible modelling environment written in Julia [@julia] for modelling the coupled interactions between ocean biogeochemistry, carbonate chemistry, and physics.

> ``OceanBioME.jl`` can be used as a stand-alone box model, or integrated into ``Oceananigans.jl`` [@Oceananigans] simulations of ocean dynamics in one-, two-, or three-dimensions.

Reviewer: ```suggestion
``OceanBioME.jl`` can be used as a stand-alone box model, or integrated into ``Oceananigans.jl`` [@Oceananigans] simulations of ocean dynamics in one, two, or three dimensions.
```

I don't think hyphens are needed since here the numbers aren't being compound modifiers, right?
Hyphens would be appropriate if, e.g., the sentence read "in one-, two-, and three-dimensional problems".

\autoref{fig2} shows a simple column model with an OCDR intervention (macroalgae growth) added after a warm-up period, which increases the carbon export of the system.

![Here we show the results of a 1D model, forced by idealised light and mixing, which qualitatively reproduces the biogeochemical cycles in the North Atlantic.
We then add kelp (500 frond / m² in the top 50 m of water) in December of the 2ⁿᵈ year (black vertical line) which causes an increase in air-sea carbon dioxide exchange and sinking export, as well as a change in the phytoplankton growth cycle.
Figure made with `Makie.jl` [@makie]. \label{fig2}](column_example.png)

The biologically active particles built into ``OceanBioME.jl`` are particularly useful for OCDR applications.
Accurate carbon accounting is essential for assessing the effectiveness of OCDR strategies.
Biologically active particles can be used to track carbon from a particular source while accounting for interactions with its surroundings.
Biologically active particles can also be used to model OCDR deployment strategies including seaweed cultivation, alkalinity enhancement, and marine biomass regeneration.

> ``OceanBioME.jl`` currently includes an extended version of the sugar kelp model presented by @broch:2012 as an example of the utility and implementation of these features.

Author: This section is kind of a statement of need, plus maybe the final paragraph. but I'm not really sure how to separate it fully into it's own section

> ``OceanBioME.jl`` was designed specifically to study ocean carbon dioxide removal (OCDR) strategies.
> Assessing the effectiveness and impacts of OCDR is challenging due to the complexities of the interactions between the biological, chemical, and physical processes involved in the carbon cycle.
> Moreover, field trials of OCDR interventions are generally small-scale and targeted, while the intervention required to have a climate-scale impact is regional or global.
> We have built ``OceanBioME.jl`` to meet these challenges by creating tools that provide a modular interface to the different components within the ocean modelling framework provided by ``Oceananigans.jl``.
> This allows easy access to a suite of biogeochemical models ranging from simple idealized to full-complexity models.
> \autoref{fig2} shows a simple column model with an OCDR intervention (macroalgae growth) added after a warm-up period, which increases the carbon export of the system.
> 
> ![Here we show the results of a 1D model, forced by idealised light and mixing, which qualitatively reproduces the biogeochemical cycles in the North Atlantic.
> We then add kelp (500 frond / m² in the top 50 m of water) in December of the 2ⁿᵈ year (black vertical line) which causes an increase in air-sea carbon dioxide exchange and sinking export, as well as a change in the phytoplankton growth cycle.
> Figure made with `Makie.jl` [@makie]. \label{fig2}](column_example.png)
> 
> The biologically active particles built into ``OceanBioME.jl`` are particularly useful for OCDR applications.
> Accurate carbon accounting is essential for assessing the effectiveness of OCDR strategies.
> Biologically active particles can be used to track carbon from a particular source while accounting for interactions with its surroundings.
> Biologically active particles can also be used to model OCDR deployment strategies including seaweed cultivation, alkalinity enhancement, and marine biomass regeneration.
> ``OceanBioME.jl`` currently includes an extended version of the sugar kelp model presented by @broch:2012 as an example of the utility and implementation of these features.
> 
> The implementation of OceanBioME.jl models allows for seamless integration with data assimilation packages, such as ``EnsembleKalmanProcesses.jl`` [@ekp]. This feature facilitates rapid calibration of model parameters, providing a powerful utility for integrating observations and models, with the potential to improve model skill and identify key sources of uncertainty.
> 
> A key metric for the validity of biogeochemical systems is the conservation of elements such as carbon and nitrogen in the system.
> We therefore continuously test the implemented models in a variety of simple scenarios (i.e. isolated, with/without air-sea flux, with/without sediment) to ensure that conservation conditions are met, and we will continue to add tests for any new models.
> Additionally, we check ``OceanBioME.jl`` utilities through standard tests such as comparison to analytical solutions for light attenuation, and conservation of tracers for active particle exudation and sinking.
> 
> Finally, this software is currently facilitating multiple research projects into ocean CDR which would have been significantly harder with other solutions.
> For example, Chen (In prep.) is using the active particle coupling provided to investigate the effects of location and planting density of kelp in the open ocean on their carbon drawdown effect, as in the example above.
> Additionally, Strong-Wright (In prep.) is using the coupling of both the biogeochemistry and easy interface to couple the physics to study flow interactions with a fully resolved giant kelp forest model including the effects on nutrient transport and distribution.

Author: ```suggestion

The implementation of OceanBioME.jl models allows for seamless integration with data assimilation packages, such as ``EnsembleKalmanProcesses.jl`` [@ekp]. This feature facilitates rapid calibration of model parameters, providing a powerful utility for integrating observations and models, with the potential to improve model skill and identify key sources of uncertainty.

A key metric for the validity of biogeochemical systems is the conservation of elements such as carbon and nitrogen in the system.
We therefore continuously test the implemented models in a variety of simple scenarios (i.e. isolated, with/without air-sea flux, with/without sediment) to ensure that conservation conditions are met, and we will continue to add tests for any new models.
Additionally, we check ``OceanBioME.jl`` utilities through standard tests such as comparison to analytical solutions for light attenuation, and conservation of tracers for active particle exudation and sinking.

# Statement of need

``OceanBioME.jl`` was designed specifically to study ocean carbon dioxide removal (OCDR) strategies.
Assessing the effectiveness and impacts of OCDR is challenging due to the complexities of the interactions between the biological, chemical, and physical processes involved in the carbon cycle.
Moreover, field trials of OCDR interventions are generally small-scale and targeted, while the intervention required to have a climate-scale impact is regional or global.
We have built ``OceanBioME.jl`` to meet these challenges by creating tools that provide a modular interface to the different components within the ocean modelling framework provided by ``Oceananigans.jl``.
This allows easy access to a suite of biogeochemical models ranging from simple idealized to full-complexity models.
\autoref{fig2} shows a simple column model with an OCDR intervention (macroalgae growth) added after a warm-up period, which increases the carbon export of the system.

![Here we show the results of a 1D model, forced by idealised light and mixing, which qualitatively reproduces the biogeochemical cycles in the North Atlantic.
We then add kelp (500 frond / m² in the top 50 m of water) in December of the 2ⁿᵈ year (black vertical line) which causes an increase in air-sea carbon dioxide exchange and sinking export, as well as a change in the phytoplankton growth cycle.
Figure made with `Makie.jl` [@makie]. \label{fig2}](column_example.png)

The biologically active particles built into ``OceanBioME.jl`` are particularly useful for OCDR applications.
Accurate carbon accounting is essential for assessing the effectiveness of OCDR strategies.
Biologically active particles can be used to track carbon from a particular source while accounting for interactions with its surroundings.
Biologically active particles can also be used to model OCDR deployment strategies including seaweed cultivation, alkalinity enhancement, and marine biomass regeneration.
``OceanBioME.jl`` currently includes an extended version of the sugar kelp model presented by @broch:2012 as an example of the utility and implementation of these features.

Finally, this software is currently facilitating multiple research projects into ocean CDR which would have been significantly harder with other solutions.
For example, Chen (In prep.) is using the active particle coupling provided to investigate the effects of location and planting density of kelp in the open ocean on their carbon drawdown effect, as in the example above.
Additionally, Strong-Wright (In prep.) is using the coupling of both the biogeochemistry and easy interface to couple the physics to study flow interactions with a fully resolved giant kelp forest model including the effects on nutrient transport and distribution.
```
Perhaps this works?

Reviewer: From JOSS docs

> A Statement of need section that clearly illustrates the research purpose of the software and places it in the context of related work.

The section above doesn't deal with the "and places it in the context of related work." Also, I'm not sure but you are suggesting to have the "Statement of need" section somewhere in the middle of the paper? I was thinking of the Statement of need section appearing straight after the Summary?

I think these three sentences

> ``OceanBioME.jl`` was designed specifically to study ocean carbon dioxide removal (OCDR) strategies. Assessing the effectiveness and impacts of OCDR is challenging due to the complexities of the interactions between the biological, chemical, and physical processes involved in the carbon cycle. Moreover, field trials of OCDR interventions are generally small-scale and targeted, while the intervention required to have a climate-scale impact is regional or global.

cover the "clearly illustrates the research purpose of the software" but don't really address the "places it in the context of related work".

Author: It's not very clear from the comment but I was thinking of putting it at the end? So we describe the code and what it can do, then in a new section do the statement of need.

It might be better if we try and write the statement of need, and then cut repetition out of the original?

Reviewer: I think so. I think a statement of need at the top and the a paragraph at the end bridging why OceanBioME.jl achieves what the Statement of need was arguing for? That's what you mean? I like that.

Reviewer: @jagoosw we haven't done this? is the JOSS waiting for this to move on with the review? 🤷‍♀️

Author: Oh yeah sorry I've been marking exams for the last two weeks so haven't done this yet

Author: I think JOSS is mainly waiting for reviewers to be found

Reviewer: When we fix it we should post `@editorialbot generate pdf` in the JOSS pre-review repo so that they get a new version of the pdf, as they may be waiting for this as well.

As a result, ``OceanBioME.jl`` and ``Oceananigans.jl`` can be used to simulate the biogeochemical response across an enormous range of scales: from surface boundary layer turbulence at the meter scale to eddying global ocean simulations at the planetary scale, and on computational systems ranging from laptops to supercomputers.
An example of a problem involving small-scale flow features is shown in \autoref{fig1}.
``OceanBioME.jl`` leverages Julia's multiple dispatch and effective inline capabilities to fuse its computations directly into existing ``Oceananigans.jl`` kernels, thus maintaining ``Oceananigans.jl``'s bespoke performance, memory- and cost-efficiency on GPUs in ``OceanBioME.jl``-augmented simulations.

![Here we replicate the Eady problem where a background buoyancy gradient and corresponding thermal wind generate a sub-mesoscale eddy, roughly following the setup of Taylor (2016).
To this physical setup, we added a medium complexity (9 tracers) biogeochemical model, some of which are shown above.
On top of this, we added particles modelling the growth of sugar kelp which are free-floating and advected by the flow, and carbon dioxide exchange from the air.

> Thanks to Julia's speed and efficiency the above model (1 km × 1 km × 100 m with 64 × 64 × 16 grid points) took about 30 minutes of computing time to simulate 10 days of evolution in about on an Nvidia P100 GPU. Figure made with `Makie` [@makie]. \label{fig1}](eady_example.png)

Reviewer: ```suggestion
Thanks to Julia's speed and efficiency the above model (1 km × 1 km × 100 m with 64 × 64 × 16 grid points) took about 30 minutes of computing time to simulate 10 days of evolution in about on an Nvidia P100 GPU. Figure made with `Makie.jl` [@makie]. \label{fig1}](eady_example.png)
```

``OceanBioME.jl`` is built with a highly modular design that allows user control and customization.
There are three distinct module types implemented in ``OceanBioME.jl``:

- First, tracer-based ecosystem modules are formulated as a set of coupled ordinary differential equations.
These equations can be solved by ``OceanBioME.jl`` as box models, which is particularly useful for testing.
The same modules can be integrated by ``Oceananigans.jl`` to provide tracer-based ecosystem models.

> - Second, boundary modules contain sets of equations which provide information at the top and bottom of the ocean.

Reviewer 2: Should it be better to point out explicitly the `Boundaries` module (for better package navigation)? (something along the lines of ``Second, the `Boundaries` module contains the sets of equations which provide information at the top and bottom of the ocean. For example, the GasExchange submodule .... ''

For example, air-sea gas exchange modules calculate the flux of carbon dioxide and oxygen at the sea surface, while sediment modules calculate fluxes of carbon and oxygen at the seafloor.

> - The third module type is "biologically active" particles.

Reviewer 2: is the module in question here the `Individuals` (`SLatissima`) as opposed to the tracer-based ecosystems which are in `AdvectedPopulations` (described in the first point of this bullet list) or `Particles`?

In the former case, wouldn't it make sense to rewrite this list by separating out the boundary conditions and highlighting that there are two `model` types (individuals based on lagrangian particle tracking and advected populations based on species transport)

Disregard if I got it wrong...

Author: I think this makes sense so will have a go at rewriting that bit now

These consist of individual-based models which are solved along particle paths and can be coupled with the tracer-based modules and physics from ``Oceananigans.jl``.
The biologically active particles can be advected by the currents, and/or they can move according to prescribed dynamics.
For example, migrating zooplankton or fish can be modelled with biologically active particles and ``OceanBioME.jl`` allows these to interact with tracer-based components such as phytoplankton or detritus.

We provide a simple framework and utilities (such as light attenuation integration) to build the necessary components of biogeochemical models.

> With the provided models, currently a simple Nutrient-Phytoplankton-Zooplankton-Detritus [@npzd] model, an intermediate complexity model, LOBSTER [@lobster], and a higher complexity model, PISCES [@pisces]. 

Reviewer 2: I couldn't find the `PISCES` implementation. Is it something that you are planning to implement?

Reviewer: True!

We have set up a straightforward "plug and play" framework to add additional tracers such as carbonate and oxygen chemistry systems, and additional forcing.
Additionally, we have implemented comprehensive air-sea flux models [e.g. @wanninkhof:1992] and sediment models [e.g. @soetaert:2000] which can easily be applied to tracers in the models.
We focus on the simulation of idealized sub-mesoscale systems, but this flexible framework allows users to model problems of any scale.
This framework is made possible by our contributions to ``Oceananigans.jl``, adding a streamlined user interface to swap biogeochemical models with no modification to other model configurations.

> This interface also facilitates rapid prototyping, as models can be coded to be in a much more accessible way, which is not possible with existing biogeochemical models.

Reviewer 2: maybe make it a bit stronger? 
e.g: ``as models can be implemented and swapped easily by just extending a few key functions. This flexibility and ease-of-use is unmatched in existing biogeochemical models''

``OceanBioME.jl`` was designed specifically to study ocean carbon dioxide removal (OCDR) strategies.
Assessing the effectiveness and impacts of OCDR is challenging due to the complexities of the interactions between the biological, chemical, and physical processes involved in the carbon cycle.
Moreover, field trials of OCDR interventions are generally small-scale and targeted, while the intervention required to have a climate-scale impact is regional or global.
We have built ``OceanBioME.jl`` to meet these challenges by creating tools that provide a modular interface to the different components within the ocean modelling framework provided by ``Oceananigans.jl``.
This allows easy access to a suite of biogeochemical models ranging from simple idealized to full-complexity models.
\autoref{fig2} shows a simple column model with an OCDR intervention (macroalgae growth) added after a warm-up period, which increases the carbon export of the system.

![Here we show the results of a 1D model, forced by idealised light and mixing, which qualitatively reproduces the biogeochemical cycles in the North Atlantic.

> We then add kelp (500 frond / m² in the top 50 m of water) in December of the 3ʳᵈ year (black vertical line) which causes an increase in air-sea carbon dioxide exchange and sinking export, as well as a change in the phytoplankton growth cycle.

Reviewer: ```suggestion
We then add kelp (500 frond / m² in the top 50 m of water) in December of the second year (black vertical line) which causes an increase in air-sea carbon dioxide exchange and sinking export, as well as a change in the phytoplankton growth cycle.
```

Doesn't look like 3rd from the figure at least.

> Figure made with `Makie` [@makie]. \label{fig2}](column_example.png)

Reviewer: ```suggestion
Figure made with `Makie.jl` [@makie]. \label{fig2}](column_example.png)
```

The biologically active particles built into ``OceanBioME.jl`` are particularly useful for OCDR applications.
Accurate carbon accounting is essential for assessing the effectiveness of OCDR strategies.
Biologically active particles can be used to track carbon from a particular source while accounting for interactions with its surroundings.
Biologically active particles can also be used to model OCDR deployment strategies including seaweed cultivation, alkalinity enhancement, and marine biomass regeneration.
``OceanBioME.jl`` currently includes an extended version of the sugar kelp model presented by @broch:2012 as an example of the utility and implementation of these features.

> We have formulated the models such that they are easy to use alongside data assimilation packages such as ``EnsembleKalmanProcesses.jl`` [@ekp] to calibrate their parameters.

Reviewer 2: also this can be a bit stronger, 
for example, ``The implementation of `OceanBioME.jl` allows for seamless integration with data assimilation packages, such as ``EnsembleKalmanProcesses.jl``. This feature facilitates rapid calibration of model parameters, providing a powerful....''

We plan to expand the features of ``OceanBioME.jl`` in the future and very much welcome user contributions.

Finally, this software is currently facilitating multiple research projects into ocean CDR which would have been significantly harder with other models.
For example, Chen et al. (in prep.) uses the active particle coupling provided to investigate the effects of location and planting density of kelp in the open ocean on their carbon drawdown effect, as in the example above.
Strong-Wright et al. (in prep.) uses the coupling of both the biogeochemistry and easy interface to couple the physics to study flow interactions with a fully resolved giant kelp forest model including the effects on nutrient transport and distribution.

# Examples

| Example| OceanBioME features utilised        | Code location |
|--------------|--------------|---------------|
| Sub-mesoscale eddy (\autoref{eady}) | LOBSTER biogeochemical model$^1$ with carbonate model active, CO$_2$ exchange with the air$^2$, Light attenuation$^3$, mass conserving negativity protection$^4$ | `examples/eady.jl` with resolution increased to 64x64x16|

> | Near-global (\autoref{global}) proof of concept | Light attenuation$^3$, NPZD model$^5$| [https://github.com/OceanBioME/GlobalOceanBioME.jl/releases/tag/v0.0.1](https://github.com/OceanBioME/GlobalOceanBioME.jl/releases/tag/v0.0.1) |

Reviewer: ```suggestion
| Near-global (\autoref{global}) proof of concept | Light attenuation$^3$, NPZD model$^5$| [github.com/OceanBioME/GlobalOceanBioME.jl/releases/tag/v0.0.1](https://github.com/OceanBioME/GlobalOceanBioME.jl/releases/tag/v0.0.1) |
```

| Idealised 1D model with kelp individuals (\autoref{column}) |  LOBSTER biogeochemical model$^1$ with carbonate model and variable Redfield ratio for organic components active, CO$_2$ exchange with the air$^2$, light attenuation$^3$, mass conserving negativity protection$^4$, and Saccharina Latissima (sugar kelp) model$^6$ | `paper/figures/column.jl`, similar to `examples/column.jl` and `examples/kelp.jl` |

$^1$ `LOBSTER`\
$^2$ `GasExchange`\
$^3$ `TwoBandPhotosyntheticallyActiveRadiation`\
$^4$ `ScaleNegativeTracers`\
$^5$ `NutrientPhytoplanktonZooplanktonDetritus`\
$^6$ `SLatissima`

# Acknowledgements

> We would like to thank the [Climate Modeling Alliance](https://clima.caltech.edu) team and ``Oceananigans.jl`` contributors for their fantastic project. We are also very grateful for the support and funding of the [Centre for Climate Repair, Cambridge](https://www.climaterepair.cam.ac.uk/) and the [Gordon and Betty Moore Foundation](https://www.moore.org/).

Reviewer: ```suggestion
We would like to thank the [Climate Modeling Alliance](https://clima.caltech.edu) team and ``Oceananigans.jl`` contributors for their fantastic project. We are also very grateful for the support and funding of the [Centre for Climate Repair, Cambridge](https://www.climaterepair.cam.ac.uk/) and the [Gordon and Betty Moore Foundation](https://www.moore.org/). N.C.C. acknowledges support by the Australian Research Council DECRA Fellowship DE210100749.
```

---
title: 'OceanBioME.jl: A flexible environment for modelling the coupled interactions between ocean biogeochemistry and physics'
tags:
  - julia
  - biogeochemistry
  - climate
  - ocean
  - carbon
authors:
  - name: Jago Strong-Wright
    orcid: 0000-0002-7174-5283
    corresponding: true
    affiliation: "1, 2"
  - name: John R Taylor
    affiliation: "1, 2"
  - name: Si Chen
    affiliation: "1, 2"
  - name: Gregory LeClaire Wagner
    orcid: 0000-0001-5317-2445
    affiliation: 3
  - name: Collaborators
affiliations:
 - name: Department of Applied Mathematics and Theoretical Physics, University of Cambridge, Cambridge, United Kingdom
   index: 1
 - name: Centre for Climate Repair, Cambridge, United Kingdom
   index: 2
 - name: Massachusetts Institute of Technology
   index: 3
date: 15 March 2023
bibliography: paper.bib
---

# Summary

> ``OceanBioME.jl`` is a flexible modelling environment written in Julia for modelling the coupled interactions between ocean biogeochemistry, carbonate chemistry, and physics. OceanBioME can be run as a stand-alone box model, or coupled with ``Oceananigans.jl`` [@Oceananigans] to run as a 1D column model or using 2D or 3D physics. OceanBioME uses Oceananigans for tracer and particle advection. This means that the biogeochemical modules built into OceanBioME can be coupled with hydrostatic or nonhydrostatic physical models. As a result, OceanBioME and Oceananigans can be used to simulate the biogeochemical response across an enormous range of scales, from non-hydrostatic 3D turbulent flows to global-scale problems. Since OceanBioME and Oceananigans are written in Julia, they are as fast as C or FORTRAN models and can easily be run across many CPUs or GPUs. Additionally, Oceananigans was written from scratch to optimize memory efficiency and GPU performance, making it extremely cost- and energy-efficient.

Author: Fantastic, thank you!

The same modules can be integrated by ``Oceananigans.jl`` to provide tracer-based ecosystem models.

- The second module type is `Individual` "biologically active" particles.
These consist of individual-based models which are solved along particle paths and can be coupled with the tracer-based modules and physics from ``Oceananigans.jl``.
The biologically active particles can be advected by the currents, and/or they can move according to prescribed dynamics.
For example, migrating zooplankton or fish can be modelled with biologically active particles and ``OceanBioME.jl`` allows these to interact with tracer-based components such as phytoplankton or detritus.

`AdvectedPopulations` are supported by `Boundaries` modules which provide information at the top and bottom of the ocean. For example, the GasExchange submodule calculate the flux of carbon dioxide and oxygen at the sea surface, while sediment modules calculate fluxes of carbon and oxygen at the seafloor.

We provide a simple framework and utilities (such as light attenuation integration) to build the necessary components of biogeochemical models.

> With the provided models, currently a simple Nutrient-Phytoplankton-Zooplankton-Detritus [@npzd] model, and an intermediate complexity model, LOBSTER [@lobster]. 

Reviewer: This sentence reads weird... Seems like you meant to write "with the provided models ... we have ..."?

Author: Oh yeah, definitely

For example, the `GasExchange` submodule calculates the carbon dioxide and oxygen flux at the sea surface, while the `Sediments` modules calculate fluxes of carbon and oxygen at the seafloor.

We currently provide a simple Nutrient-Phytoplankton-Zooplankton-Detritus (NPZD) model [@npzd], and an intermediate complexity model, LOBSTER [@lobster], we have set up a straightforward "plug and play" framework to add additional tracers such as carbonate and oxygen chemistry systems and additional forcing. 
A key feature of this package is the easy ability to modify models or add different formulations. 
If a user wanted to implement a different model they could use the existing ones as a template and modify only a few lines of code where the ODEs are defined as functions.
Alternatively, by harnessing another Julia feature, users could `import` specific functions from the models and overload only them.
For example, if a user wanted to modify the light limitation to phytoplankton growth in the NPZD model to copy that of the LOBSTER model they could pass different parameter values to the model and write:

```julia
import OceanBioME.NPZDModel: light_limitation

> @inline light_limitation(PAR, α, μ₀) = 1 - exp(- PAR / α)

Reviewer 3: This doesn't overload `light_limitation` (there's no type annotation?)

Author: Should I change it to call it redefining?

Reviewer 3: Ah no, the code should be changed so that method extension is possible! Does this code example achieve it's intended goal? I'm not quite sure what it's trying to do. What are the "other" methods associated with the name `light_limitation`?

Reviewer 3: In terms of what to call it, this example defines a fallback for `light_limitation`. What's unclear to me is why this would achieve any purpose. The concept of modularity and extension requires _both_ introducing a novel type, and then extending methods for that specific type to achieve some different behavior than what is already implemented.

Author: Ah so I think I've a) not explained this example very well and b) its definitely bad practice so maybe we shouldn't encourage it.

Here I used the first simple example I thought of and was thinking that e.g. a user wants to experiment with how the formulation of the light limitation to phytoplankton growth affects the NPZD model. This is included in the NPZD model in this line:
```julia
growth = μ₀ * Q₁₀(T) * nutrient_limitation(N, kₙ) * light_limitation(PAR, α, μ₀ * Q₁₀(T)) * P
```
so if you could change just `light_limitation`. So they could write a script that does this at the start, then runs a model. I checked and this does work to replace the function. I checked and it does do this.

Perhaps we should come up with a better example. I don't think, as currently written, it would be straightforward to modify models "properly". I do have a solution to this that I've been working on, but its a lot of changes and I was going to suggest we all talk about it at some point.
